technical-docs/designs/infra/rollback.md

```diff
@@ -0,0 +1,218 @@
+`after release` trigger as the trigger that will start the deployment to this
+stage. Next we will enable Gates and add an arbitrary delay time of 10 minutes
+before evaluation to assure\ all are logs have been generated.
+
+![](images/conditiongate.png)
+
+Here we can use custom gates to invoke our rollback check including community
+available gates provided on the
+[Visual Studio marketplace](https://marketplace.visualstudio.com/search?term=tag%3ARelease%20Gates&target=AzureDevOps&category=All%20categories&visibilityQuery=all&sortBy=Relevance).
+In this example, we `Query Azure Monitor alerts` to check the status of our AKS
+cluster deployed using terraform and Application Insights to assure its in a
+healthy state. In my Application Insights, I have an alert that checks the
+readiness of my aks pods over an aggregated time.
+
+![](images/alertLogic.png)
+
+You're able define a variety of alerts that target services or databases using
+[Azure Alerts](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-overview_).
+In the pre-deployment gate I add my previously configured alert as my condition.
+
+![](images/azuremonitorgate.png)
+
+Finally we can add evaluation options that dictate the time between
+re-evaluation of the gate, the timeout after which the gate will fail and also
+approval after the successful gate has been triggered.
+
+### 3.3 Terraform Backend Rollback
+
+With _Azure DevOps Pipeline Environments_, we can monitor Azure Web Apps, and
+Kubernetes clusters as resources of deployments. We can also track the health
+state of workloads and services within the cluster. Build deployments are
+associated with PR git hashes and we can use this to identify a previous healthy
+state to roll back our environment to.
+
+![](images/environmentService.png)
+
+Let's navigate back to our _Rollback Check_ stage and add a task to Download
+Pipeline Artifacts. Here we can download a previous build artifact that can be
+tagged to a specific version. The build version number can be passed in from a
+variable group that tracks successful builds of your resources.
+
+![](images/downloadartifact.png)
+
+Now you have the option to run pre-validation tests to assure the
+`terraform apply` will be successful. For simplicity we run `terraform init`,
+`plan` in this stage. When configuring the terraform tasks, use your variable
+group to pass the values for the `Resource group`, `storage account` and
+`Container`. The `key` will be the previous terraform state key produced in from
+the prior build artifact. Below is an example of a YAML:
+
+```
+steps:
+- task: ms-devlabs.custom-terraform-tasks.custom-terraform-release-task.TerraformTaskV1@0
+  displayName: 'Terraform : apply'
+  inputs:
+    backendServiceArm: '<Sub_ID>'
+    backendAzureRmResourceGroupName: 'nr-aks-hlf'
+    backendAzureRmStorageAccountName: nrhype
+    backendAzureRmContainerName: nrhype
+    backendAzureRmKey: $(Previous_key_git_version)
```

Comment:
Is this the one named using the git hash?

Reply:
Yes, I updated the doc to make that more clear.